### ISSUE TITLE
Zero 3, async_inplace_copy_grad_to_fp32_buffer_from_gpu differences between code and paper.

### ISSUE THREAD

**Issue by feifeibear** [microsoft/DeepSpeed#899]
Hi, 
I am learning the implementation of `async_inplace_copy_grad_to_fp32_buffer_from_gpu` from 
https://github.com/microsoft/DeepSpeed/blob/4601885972/deepspeed/runtime/zero/stage3.py#L1570

I noticed there are some gaps between the code and the description of the paper (ZeRO-Offload: Democratizing Billion-Scale Model Training).
As shown in Figure 3 of the paper, it should be that the fp16 gradients are copying from gpu to cpu. However, in the code, it is the fp32 gradients being copied, which is 4M in this case, different from 2M mentioned in the paper.
I wonder if my understanding is correct? If so, why the code and paper are different. (for accuracy considerations? or some other reasons?)

**Comment by samyam** [comment]
@feifeibear, good catch! The reason why the code and paper are different for this piece is due to PyTorch performance limitations . Converting the fp16 tensors to fp32 tensors on the CPU with PyTorch is super slow. So instead we convert it to fp32 on the GPU before moving it to CPU so that we don't have to make that conversion on the CPU. It is faster to communicate in fp32 than to communicate in fp16 then convert it in CPU using the pytorch conversion operators.

**Comment by feifeibear** [comment]
Thanks for your reply. That makes sence.
